nlp/model/augmented_lagrangian.py

```diff
@@ -128,6 +129,21 @@ def hess(self, *args, **kwargs):
         return self.hop(*args, **kwargs)
 
 
+    def magical_step(self, x, g):
+        """Compute a magical step for the slack variables.
+
+        This step minimizes the augmented Lagrangian with respect to the slack
+        variables only for a fixed set of decision variables. Here, g is the 
+        gradient computed by self.grad at x.
+        """
+        model = self.model
+        on = model.original_n
+        m_step = np.zeros(self.n)
+        m_step[on:] = -g[on:] / self.penalty
+        m_step = projected_step(x, m_step, model.Lvar, model.Uvar)
+        return m_step
+
+
```

Comment:
The gradient of the augmented Lagrangian w.r.t. the slack variables is exactly _y - ρ(c(x) - s)_ and the Hessian is ρ times the identity matrix. Therefore, if you have the current gradient, you can compute the optimal step by just dividing the slack variable part by ρ. This was how we did it in the old Auglag. 

Having said that, you're right. We could just compute the optimal slack variables directly given the constraints at the current point. I can change the code to do it that way.